analytics/upress_reporting/helpers/download_data_aggregator.py

```diff
@@ -0,0 +1,142 @@
+import os
+import boto3
+import re
+
+from analytics.upress_reporting.models.data.download_event import DownloadEvent
+from logger import createLog
+from model import Edition, Item, Link
+from model.postgres.item import ITEM_LINKS
+from managers import DBManager
+
+# Regexes needed to parse S3 logs
+REQUEST_REGEX = r"REST.GET.OBJECT "
+# File ID includes the file name for the pdf object
+FILE_ID_REGEX = r"REST.GET.OBJECT (.+pdf\s)"
+TIMESTAMP_REGEX = r"\[.+\]"
+REFERRER_REGEX = r"https://drb-qa.nypl.org/"
+
+
+class DownloadDataAggregator:
+    """
+    Parses S3 download logs and generates list of DownloadEvents, each corresponding 
+    to a single download request.
+    """
+
+    def __init__(self, publisher, date_range):
+        self.publisher = publisher
+        self.date_range = date_range
+
+        self.s3_client = boto3.client("s3")
+        self.bucket_name = os.environ.get("DOWNLOAD_BUCKET", None)
+        self.log_path = os.environ.get("DOWNLOAD_LOG_PATH", None)
+
+        self._setup_db_manager()
+        self.logger = createLog("download_request_parser")
+
+    def pull_download_events(self):
+        '''
+        Returns list of DownloadEvents in a given reporting period.
+        '''
+        download_events = []
+
+        for date in self.date_range:
+            folder_name = date.strftime("%Y/%m/%d")
+            batch = self._load_batch(folder_name)
+            downloads_per_day = self._parse_logs(batch)
+            download_events.extend(downloads_per_day)
+        
+        self.db_manager.session.close()
```

Comment:
hmm it may be good to use a with clause for this function to ensure that the session is closed!

Reply:
Hm are you thinking something like 
```
with self.db_manager.createSession():
        download_events = []

        for date in self.date_range:
            folder_name = date.strftime("%Y/%m/%d")
            batch = self._load_batch(folder_name)
            downloads_per_day = self._parse_logs(batch)
            download_events.extend(downloads_per_day)

        return download_events
```
Otherwise I'm not sure how to do that haha

Reply:
Otherwise a try-except-finally makes sense to me (with the finally being self.db_manager.session.close())

Reply:
As discussed offline, Kyle will add enter and exit functions to the DBManager class as part of his changes!